src/snowflake/connector/connection.py

```diff
@@ -1541,13 +1543,27 @@ def is_an_error(status: QueryStatus) -> bool:
 
     def _all_async_queries_finished(self) -> bool:
         """Checks whether all async queries started by this Connection have finished executing."""
-        queries = copy.copy(
-            self._async_sfqids
-        )  # get_query_status might update _async_sfqids, let's copy the list
-        finished_async_queries = (
-            not self.is_still_running(self.get_query_status(q)) for q in queries
-        )
-        return all(finished_async_queries)
+
+        queries = list(reversed(list(self._async_sfqids.keys())))
+        if not queries:
+            return True
+        num_workers = min(self.client_prefetch_threads, len(queries))
+        found_unfinished_query = False
+
+        def check_status_helper(
+            sfq_id: str,
+        ):  # We should upgrade to using cancel_futures=True once supporting 3.9+
+            nonlocal found_unfinished_query
+            if not found_unfinished_query and self.is_still_running(
+                self.get_query_status(sfq_id)
+            ):
+                found_unfinished_query = True
+
+        with ThreadPoolExecutor(max_workers=num_workers) as tpe:
+            for sfq_id in queries:
+                tpe.submit(check_status_helper, sfq_id)
```

Comment:
I think you forgot to wait on the futures that [submit](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.submit) returns.
Could I instead suggest using [map](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map)? The trick with that helper is to cast it to a list to wait on all of the results.
Otherwise you'll need to construct a list of these futures and wait on each one of them.

Reply:
I think using `with` means it does an implicit `shutdown(wait=True)` when exiting. Since all the work it does is in the `callback`, I don't need to wait on the futures explicitly in the main thread. But maybe this is not good for code readability?

Reply:
We could get rid of `callback` and replace it with `map` and `as_completed`, and do the post-processing in the main thread 🤔

Reply:
> I think using `with` means it does an implicit `shutdown(wait=True)` when exiting. Since there is no post-processing, technically I don't need to wait on the futures explicitly in the main thread. But maybe that is not good for code readability?

I forgot about this, good catch!